Source/hydro/fourth_order.F90

```diff
@@ -56,375 +59,849 @@ subroutine states(idir, &
-                            ! limit
-                            if (dafm(i,j,k)*dafp(i,j,k) < 0.0_rt) then
-                               ! Eqs. 29, 30
-                               ar(i,j,k,n) = a(i,j,k,n) - rho*dafm(i,j,k)  ! note: typo in Eq 29
-                               al(i,j,k+1,n) = a(i,j,k,n) + rho*dafp(i,j,k)
-                            else if (abs(dafm(i,j,k)) >= 2.0_rt*abs(dafp(i,j,k))) then
-                               ! Eq. 31
-                               ar(i,j,k,n) = a(i,j,k,n) - 2.0_rt*(1.0_rt - rho)*dafp(i,j,k) - rho*dafm(i,j,k)
-                            else if (abs(dafp(i,j,k)) >= 2.0*abs(dafm(i,j,k))) then
-                               ! Eq. 32
-                               al(i,j,k+1,n) = a(i,j,k,n) + 2.0_rt*(1.0_rt - rho)*dafm(i,j,k) + rho*dafp(i,j,k)
-                            endif
-
-                         endif
-                      endif
+                else
+                   ! not supported
+                   call amrex_error("ERROR: boundary conditions not supported for +Z in states")
+                end if
 
-                   else
-                      ! if Eqs. 24 or 25 didn't hold we still may need to limit
-                      if (abs(dafm(i,j,k)) >= 2.0_rt*abs(dafp(i,j,k))) then
-                         ar(i,j,k,n) = a(i,j,k,n) - 2.0_rt*dafp(i,j,k)
-                      endif
-                      if (abs(dafp(i,j,k)) >= 2.0_rt*abs(dafm(i,j,k))) then
-                         al(i,j,k+1,n) = a(i,j,k,n) + 2.0_rt*dafm(i,j,k)
-                      endif
-                   endif
-
-                   ! apply flattening
-                   al(i,j,k+1,n) = flatn(i,j,k)*al(i,j,k+1,n) + (ONE - flatn(i,j,k))*a(i,j,k,n)
-                   ar(i,j,k,n) = flatn(i,j,k)*ar(i,j,k,n) + (ONE - flatn(i,j,k))*a(i,j,k,n)
-
-                enddo
-             enddo
-          enddo
-       endif
-    endif
+             end do
+          end do
+
+       end if
+
+    end if
 
   end subroutine states
 
 
   ! Note: pretty much all of these routines below assume that dx(1) = dx(2) = dx(3)
+  pure function compute_laplacian(i, j, k, n, &
+                                  a, a_lo, a_hi, nc, &
+                                  domlo, domhi) result (lap)
+
+    use prob_params_module, only : physbc_lo, physbc_hi, Interior
+    implicit none
+
+    integer, intent(in) :: i, j, k, n
+    integer, intent(in) :: a_lo(3), a_hi(3)
+    integer, intent(in) :: nc
+    real(rt), intent(in) :: a(a_lo(1):a_hi(1), a_lo(2):a_hi(2), a_lo(3):a_hi(3), nc)
+    integer, intent(in) :: domlo(3), domhi(3)
+
+    real(rt) :: lapx, lapy, lapz
+    real(rt) :: lap
+
+    lapx = ZERO
+    lapy = ZERO
+    lapz = ZERO
+
+    ! we use 2nd-order accurate one-sided stencils at the physical boundaries
+
+    if (i == domlo(1) .and. physbc_lo(1) /= Interior) then
+       lapx = 2.0_rt*a(i,j,k,n) - 5.0_rt*a(i+1,j,k,n) + 4.0_rt*a(i+2,j,k,n) - a(i+3,j,k,n)
```

Comment:
This one-sided second-order stencil looks correct to me, but I notice that McCorquodale and Colella suggest a different approach: if I read it correctly, set the Laplacian here to just be the Laplacian at domlo(1) + 1. I like what you've done more. Any thoughts here? Can you add a comment explicitly saying why you did it differently than the paper so we don't trip up ourselves up on this later?

Reply:
I basically didn't like MC's approach here.  I say so in the SDC paper.  I think my approach is better, but I agree, I should add a comment noting the difference.